server/sonar-server/src/main/java/org/sonar/server/es/BulkIndexer.java

```diff
@@ -63,35 +65,99 @@
     this.client = client;
     this.indexName = indexName;
     this.progress = new ProgressLogger(format("Progress[BulkIndexer[%s]]", indexName), counter, LOGGER)
       .setPluralLabel("requests");
-
-    // see https://jira.sonarsource.com/browse/SONAR-8075
-    this.concurrentRequests = Math.max(1, Runtime.getRuntime().availableProcessors() / 5);
-    this.semaphore = new Semaphore(concurrentRequests);
   }
 
   public enum Size {
     /** Use this size for a limited number of documents. */
     REGULAR,
 
     /** Use this size for initial indexing and if you expect unusual huge numbers of documents. */
-    LARGE;
+    LARGE {
+
+      @Override
+      void beforeStart(BulkIndexer bulkIndexer) {
+        bulkIndexer.setLargeInitialSettings(Maps.newHashMap());
+        Map<String, Object> bulkSettings = Maps.newHashMap();
+        GetSettingsResponse settingsResp = bulkIndexer.client.nativeClient().admin().indices().prepareGetSettings(bulkIndexer.indexName).get();
+
+        // deactivate replicas
+        int initialReplicas = Integer.parseInt(settingsResp.getSetting(bulkIndexer.indexName, IndexMetaData.SETTING_NUMBER_OF_REPLICAS));
+        if (initialReplicas > 0) {
+          bulkIndexer.getLargeInitialSettings().put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, initialReplicas);
+          bulkSettings.put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0);
+        }
+
+        // deactivate periodical refresh
+        String refreshInterval = settingsResp.getSetting(bulkIndexer.indexName, REFRESH_INTERVAL_SETTING);
+        bulkIndexer.getLargeInitialSettings().put(REFRESH_INTERVAL_SETTING, refreshInterval);
+        bulkSettings.put(REFRESH_INTERVAL_SETTING, "-1");
+
+        updateSettings(bulkIndexer, bulkSettings);
+      }
+
+      @Override
+      void afterStop(BulkIndexer bulkIndexer) {
+        // optimize lucene segments and revert index settings
+        // Optimization must be done before re-applying replicas:
+        // http://www.elasticsearch.org/blog/performance-considerations-elasticsearch-indexing/
+        bulkIndexer.client.prepareForceMerge(bulkIndexer.indexName).get();
+
+        updateSettings(bulkIndexer, bulkIndexer.getLargeInitialSettings());
+      }
+
+      private void updateSettings(BulkIndexer bulkIndexer, Map<String, Object> settings) {
+        UpdateSettingsRequestBuilder req = bulkIndexer.client.nativeClient().admin().indices().prepareUpdateSettings(bulkIndexer.indexName);
+        req.setSettings(settings);
+        req.get();
+      }
+    };
+
+    void beforeStart(BulkIndexer bulkIndexer) {
+      // can be overwritten
+    }
+
+    void afterStop(BulkIndexer bulkIndexer) {
+      // can be overwritten
+    }
+  }
+
+  /** @see https://jira.sonarsource.com/browse/SONAR-8075 */
+  @VisibleForTesting
+  static int getConcurrentRequests(BulkIndexer bulkIndexer) {
+    int simultaniousRequests = Math.max(1, bulkIndexer.getProcesses() / getNumberOfShards(bulkIndexer));
```

Comment:
Still checking the number of shards ? As discussed the old hardcoded value `5` should be kept.